File: python-algorithm/ds_4_linkedlist/linkedlist_2_inplace_operation/L24-m.py
#--coding:utf-8--

# 这道题是三个指针
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        dummy = ListNode(0, head)
        back = dummy
        cur = head
        # 三指针
        while cur and cur.next:
            fut = cur.next
            cur.next = fut.next
            fut.next = cur
            back.next = fut
            back = cur
            cur = cur.next
        return dummy.next

File: python-algorithm/ds_4_linkedlist/linkedlist_2_inplace_operation/L24-m.cpp

//leetcode 24 交换链表节点

#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val ;
    ListNode* next ;
    ListNode(int x): val(x) , next(NULL) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head==NULL||head->next == NULL)
            return head;
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* pre = dummy;
        ListNode* back = head;
        while(head && head->next){
            back = head->next;
            pre->next = back;
            head->next = back->next;
            back->next = head;
            pre = head;
            head = pre->next;
        }
        ListNode* res = dummy->next;
        delete dummy;
        return res;
    }
};


int main(){
    return 0 ;
}

//练习leetcode 25 147(链表插入排序) 148（对链表排序）